File: Scripts/doBrainTissueSegmentation.R
library( ANTsR )
library( ANTsRNet )
library( keras )

args <- commandArgs( trailingOnly = TRUE )

if( length( args ) != 4 )
  {
  helpMessage <- paste0( "Usage:  Rscript doHippoEcSegmentation.R",
    " inputFile inputMaskFile outputFilePrefix reorientationTemplate\n" )
  stop( helpMessage )
  } else {
  inputFileName <- args[1]
  inputMaskFileName <- args[2]
  outputFilePrefix <- args [3]
  reorientTemplateFileName <- args[4]
  }

classes <- c( "Background", "Csf", "GrayMatter", "WhiteMatter",
  "DeepGrayMatter", "BrainStem", "Cerebellum"  )
numberOfClassificationLabels <- length( classes )
resampledImageSize <- c( 112, 160, 112 )
segmentationLabels <- seq_len( numberOfClassificationLabels ) - 1
numberOfFiltersAtBaseLayer <- 16

imageMods <- c( "T1" )
channelSize <- length( imageMods )

cat( "Reading reorientation template", reorientTemplateFileName )
startTime <- Sys.time()
reorientTemplate <- antsImageRead( reorientTemplateFileName )
endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

unetModel <- createUnetModel3D( c( resampledImageSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, dropoutRate = 0.0,
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = numberOfFiltersAtBaseLayer,
  convolutionKernelSize = c( 3, 3, 3 ), deconvolutionKernelSize = c( 2, 2, 2 ),
  weightDecay = 1e-5 )

cat( "Loading weights file" )
startTime <- Sys.time()
weightsFileName <- getPretrainedNetwork( "brainSegmentation" )
load_model_weights_hdf5( unetModel, filepath = weightsFileName )
endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

# Process input

startTimeTotal <- Sys.time()

cat( "Reading ", inputFileName )
startTime <- Sys.time()
image <- antsImageRead( inputFileName, dimension = 3 )
mask <- antsImageRead( inputMaskFileName, dimension = 3 )
endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

cat( "Normalizing to template and cropping to mask." )
startTime <- Sys.time()
centerOfMassTemplate <- getCenterOfMass( reorientTemplate )
centerOfMassImage <- getCenterOfMass( image )
xfrm <- createAntsrTransform( type = "Euler3DTransform",
  center = centerOfMassTemplate,
  translation = centerOfMassImage - centerOfMassTemplate )
warpedImage <- applyAntsrTransformToImage( xfrm, image, reorientTemplate )
warpedMask <- applyAntsrTransformToImage( xfrm, mask, reorientTemplate,
  interpolation = "nearestNeighbor" )
warpedMask <- iMath( warpedMask, "MD", 3 )
warpedCroppedImage <- cropImage( warpedImage, warpedMask, 1 )
originalCroppedSize <- dim( warpedCroppedImage )
warpedCroppedImage <- resampleImage( warpedCroppedImage,
  resampledImageSize, useVoxels = TRUE )
endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

batchX <- array( data = as.array( warpedCroppedImage ),
  dim = c( 1, resampledImageSize, channelSize ) )
batchX <- ( batchX - mean( batchX ) ) / sd( batchX )

cat( "Prediction and decoding" )
startTime <- Sys.time()
predictedData <- unetModel %>% predict( batchX, verbose = 0 )
probabilityImagesArray <- decodeUnet( predictedData, warpedCroppedImage )
endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( " (elapsed time:", elapsedTime, "seconds)\n" )

cat( "Renormalize to native space" )
startTime <- Sys.time()

zeroArray <- array( data = 0, dim = dim( warpedImage ) )
zeroImage <- as.antsImage( zeroArray, reference = warpedImage )

probabilityImages <- list()
for( i in seq_len( numberOfClassificationLabels ) )
  {
  probabilityImageTmp <- resampleImage( probabilityImagesArray[[1]][[i]],
    originalCroppedSize, useVoxels = TRUE )
  probabilityImageTmp <- decropImage( probabilityImageTmp, zeroImage )
  probabilityImages[[i]] <- applyAntsrTransformToImage( invertAntsrTransform( xfrm ),
    probabilityImageTmp, image )
  }

endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

cat( "Writing", outputFilePrefix )
startTime <- Sys.time()

probabilityImageFiles <- c()
for( i in seq_len( numberOfClassificationLabels - 1 ) )
  {
  probabilityImageFiles[i] <- paste0( outputFilePrefix, classes[i+1], ".nii.gz" )
  antsImageWrite( probabilityImages[[i+1]], probabilityImageFiles[i] )
  }

probabilityImagesMatrix <- imagesToMatrix( probabilityImageFiles, mask )
segmentationVector <- apply( probabilityImagesMatrix, FUN = which.max, MARGIN = 2 )
segmentationImage <- makeImage( mask, segmentationVector )
antsImageWrite( segmentationImage, paste0( outputFilePrefix, "Segmentation.nii.gz" ) )

endTime <- Sys.time()
elapsedTime <- endTime - startTime
cat( "  (elapsed time:", elapsedTime, "seconds)\n" )

endTimeTotal <- Sys.time()
elapsedTimeTotal <- endTimeTotal - startTimeTotal
cat( "\nTotal elapsed time:", elapsedTimeTotal, "seconds\n\n" )

File: Scripts/doBrainTissueSegmentation.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
# vi: set ft=python sts=4 ts=4 sw=4 et

import os
import sys
import time
import numpy as np
import keras

import ants
import antspynet

args = sys.argv

if len(args) != 5:
    help_message = ("Usage:  python doBrainExtraction.py" +
        " inputFile inputMaskFile outputFilePrefix reorientationTemplate")
    raise AttributeError(help_message)
else:
    input_file_name = args[1]
    input_mask_file_name = args[2]
    output_file_name_prefix = args[3]
    reorient_template_file_name = args[4]

classes = ("Background", "Csf", "GrayMatter", "WhiteMatter",
  "DeepGrayMatter", "BrainStem", "Cerebellum")
number_of_classification_labels = len(classes)
segmentation_labels = list(range(0, number_of_classification_labels))
resampled_image_size = (112, 160, 112)

image_mods = ["T1"]
channel_size = len(image_mods)

print("Reading reorientation template " + reorient_template_file_name)
start_time = time.time()
reorient_template = ants.image_read(reorient_template_file_name)
end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")


unet_model = antspynet.create_unet_model_3d((*resampled_image_size, channel_size),
  number_of_outputs = number_of_classification_labels,
  number_of_layers = 4,
  number_of_filters_at_base_layer = 16,
  dropout_rate = 0.0,
  convolution_kernel_size = (3, 3, 3),
  deconvolution_kernel_size = (2, 2, 2),
  weight_decay = 1e-5 )

print( "Loading weights file" )
start_time = time.time()
weights_file_name = "./brainSegmentationWeights.h5"

if not os.path.exists(weights_file_name):
    weights_file_name = antspynet.get_pretrained_network("brainExtraction", weights_file_name)

unet_model.load_weights(weights_file_name)
end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")

start_time_total = time.time()

print( "Reading ", input_file_name )
start_time = time.time()
image = ants.image_read(input_file_name)
mask = ants.image_read( input_mask_file_name)
end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")

print( "Normalizing to template" )
start_time = time.time()
center_of_mass_template = ants.get_center_of_mass(reorient_template)
center_of_mass_image = ants.get_center_of_mass(image)
translation = np.asarray(center_of_mass_image) - np.asarray(center_of_mass_template)
xfrm = ants.create_ants_transform(transform_type="Euler3DTransform",
  center=np.asarray(center_of_mass_template),
  translation=translation)
warped_image = ants.apply_ants_transform_to_image(xfrm, image,
  reorient_template, interpolation='linear')
warped_mask = ants.apply_ants_transform_to_image(xfrm, mask, reorient_template,
  interpolation='linear' )
warped_mask = ants.threshold_image(warped_mask, 0.4999, 1.0001, 1, 0)
warped_mask = ants.iMath(warped_mask, "MD", 3)
warped_cropped_image = ants.crop_image(warped_image, warped_mask, 1)
original_cropped_size = warped_cropped_image.shape
warped_cropped_image = ants.resample_image( warped_cropped_image,
  resampled_image_size, use_voxels = True )
end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")

batchX = np.expand_dims(warped_cropped_image.numpy(), axis=0)
batchX = np.expand_dims(batchX, axis=-1)
batchX = (batchX - batchX.mean()) / batchX.std()

print("Prediction and decoding")
start_time = time.time()
predicted_data = unet_model.predict(batchX, verbose=0)

origin_cropped = warped_cropped_image.origin
spacing_cropped = warped_cropped_image.spacing
direction_cropped = warped_cropped_image.direction

probability_images_array = list()
for i in range(number_of_classification_labels):
    probability_images_array.append(
       ants.from_numpy(np.squeeze(predicted_data[0, :, :, :, i]),
         origin=origin_cropped, spacing=spacing_cropped,
         direction=direction_cropped))

end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")

print("Renormalize to native space")
start_time = time.time()

zeros = np.zeros(warped_image.shape)
zeros_image = ants.from_numpy( zeros, origin=warped_image.origin,
  spacing=warped_image.spacing, direction=warped_image.direction)

for i in range(number_of_classification_labels):
    probability_image = ants.resample_image(probability_images_array[i],
      original_cropped_size, use_voxels=True, interp_type=0)
    probability_image = ants.decrop_image(probability_image, zeros_image)
    probability_images_array[i] = ants.apply_ants_transform_to_image(
      ants.invert_ants_transform(xfrm), probability_image, image)

end_time = time.time()
elapsed_time = end_time - start_time
print("  (elapsed time: ", elapsed_time, " seconds)")

for i in range(1, number_of_classification_labels):
    print("Writing", classes[i])
    start_time = time.time()
    ants.image_write(probability_images_array[i],
      output_file_name_prefix + classes[i] + "Segmentation.nii.gz")
    end_time = time.time()
    elapsed_time = end_time - start_time
    print("  (elapsed time: ", elapsed_time, " seconds)")

probability_images_matrix = ants.image_list_to_matrix(probability_images_array, mask)
segmentation_vector = np.argmax(probability_images_matrix, axis=0)
segmentation_image = ants.make_image(mask, segmentation_vector)
ants.image_write(segmentation_image, output_file_name_prefix + "Segmentation.nii.gz")

end_time_total = time.time()
elapsed_time_total = end_time_total - start_time_total
print("Total elapsed time: ", elapsed_time_total, "seconds")